# Solutions — Supermarket Sales Practice

Answer key for all 80 questions.

Try each question yourself first, then check here. If your code looks different but gives the same result, that is fine — there is usually more than one correct way to write it.

Every solution in this file has been run and tested, so the notebook runs top to bottom without errors.

---

## Setup — run this cell first

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", 30)

URL = ("https://raw.githubusercontent.com/sushantag9/"
       "Supermarket-Sales-Data-Analysis/master/supermarket_sales%20-%20Sheet1.csv")
df = pd.read_csv(URL)

print("shape:", df.shape)
df.head()

shape: (1000, 17)


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


---

# Part A — NumPy (30 questions)

First run the cell below to create the arrays. All the NumPy questions use them.

In [ ]:
cols = ["Unit price", "Quantity", "Tax 5%", "Total", "cogs", "gross income", "Rating"]
arr = df[cols].to_numpy()

total = df["Total"].to_numpy()
rating = df["Rating"].to_numpy()
qty = df["Quantity"].to_numpy()

print("arr shape:", arr.shape)

arr shape: (1000, 7)


**Q1.** Print the shape, number of dimensions, size and data type of `arr`.

In [ ]:
print( arr.shape)
print( arr.ndim)
print( arr.size)
print( arr.dtype)

(1000, 7)
2
7000
float64


**Q2.** Create an array of numbers from 0 to 11 and reshape it into a 3 by 4 matrix.

In [ ]:
m = np.arange(12).reshape(3, 4)
print(m)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


**Q3.** Create a 3 by 3 array of all zeros, then a 3 by 3 array of all ones.

In [ ]:
print(np.zeros((3, 3)))
print()
print(np.ones((3, 3)))

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


**Q4.** Use `np.linspace` to create 5 equally spaced numbers between 0 and 100.

In [ ]:
print(np.linspace(0, 100, 5))

[  0.  25.  50.  75. 100.]


**Q5.** Print the first 5 rows of `arr`.

In [ ]:
print(arr[:5])

[[ 74.69     7.      26.1415 548.9715 522.83    26.1415   9.1   ]
 [ 15.28     5.       3.82    80.22    76.4      3.82     9.6   ]
 [ 46.33     7.      16.2155 340.5255 324.31    16.2155   7.4   ]
 [ 58.22     8.      23.288  489.048  465.76    23.288    8.4   ]
 [ 86.31     7.      30.2085 634.3785 604.17    30.2085   5.3   ]]


**Q6.** Print the first 10 values of the Unit price column. It is column number 0.

In [ ]:
print(arr[:10, 0])

[74.69 15.28 46.33 58.22 86.31 85.39 68.84 73.56 36.26 54.84]


**Q7.** Print the last 3 rows of `arr`.

In [ ]:
print(arr[-3:])

[[ 31.84    1.      1.592  33.432  31.84    1.592   7.7  ]
 [ 65.82    1.      3.291  69.111  65.82    3.291   4.1  ]
 [ 88.34    7.     30.919 649.299 618.38   30.919   6.6  ]]


**Q8.** Print every 100th value of the `total` array using slicing.

In [ ]:
print(total[::100])

[548.9715  51.1455 120.645  209.622  248.409   77.6685  87.234  507.444
 144.963  423.738 ]


The pattern is `start : stop : step`. Leaving start and stop empty means the whole array.

**Q9.** Print the values of `total` at positions 0, 10, 50 and 999.

In [ ]:
print(total[[0, 10, 50, 999]])

[548.9715  60.816  671.79   649.299 ]


Two sets of brackets. The inner one is the list of positions, the outer one does the indexing.

**Q10.** Reverse the `rating` array and print its first 5 values.

In [ ]:
rev = rating[::-1]
print(rev[:5])

[6.6 4.1 7.7 4.4 6.2]


**Q11.** Print the mean, median, minimum and maximum of the `total` array.

In [ ]:
print(total.mean())
print(np.median(total))
print(total.min())
print(total.max())

322.966749
253.848
10.6785
1042.65


**Q12.** Print the standard deviation of the `total` array.

In [ ]:
print( total.std())

245.76236168237705


**Q13.** Print the 25th, 50th and 75th percentile of the `rating` array.

In [ ]:
print(np.percentile(rating, [25, 50, 75]))

[5.5 7.  8.5]


**Q14.** Find the mean of every column of `arr` using axis=0.

In [ ]:
print(arr.mean(axis=0))

[ 55.67213    5.51      15.379369 322.966749 307.58738   15.379369
   6.9727  ]


axis=0 goes down the rows, so you get one value per column.

**Q15.** Find the sum of every row of `arr` using axis=1. Print the first 5 results.

In [ ]:
row_sums = arr.sum(axis=1)
print(row_sums[:5])

[1214.8745  194.14    757.9965 1076.004  1397.5755]


axis=1 goes across the columns, so you get one value per row.

**Q16.** Find how many sales have a Total greater than 500.

In [ ]:

print((total > 500).sum())

227


`total > 500` gives an array of True and False. True counts as 1, so `.sum()` gives the number of True values.

**Q17.** Print that same number as a percentage of all 1000 sales.

In [ ]:
count = (total > 500).sum()
print(round(count / len(total) * 100, 2), "%")

22.7 %


**Q18.** Using boolean masking, get all rating values above 8 and print how many there are.

In [ ]:
high = rating[rating > 8]
print(len(high))

308


**Q19.** Print the average of those ratings above 8.

In [ ]:
high = rating[rating > 8]
print(round(high.mean(), 2))

9.04


**Q20.** Find how many sales have a Total above 300 and a Rating above 7 at the same time.

In [ ]:
print(((total > 300) & (rating > 7)).sum())

203


Use `&` for and, not the word `and`. Each condition needs its own brackets.

**Q21.** Use `np.where` to make an array that holds 1 when Total is above 400 and 0 otherwise. Print its sum.

In [ ]:
flags = np.where(total > 400, 1, 0)
print(flags.sum())

320


**Q22.** Use `np.where` to label each rating as 'Good' when it is above 7 and 'Not Good' otherwise. Print the first 10 labels.

In [ ]:
labels = np.where(rating > 7, "Good", "Not Good")
print(labels[:10])

['Good' 'Good' 'Good' 'Good' 'Not Good' 'Not Good' 'Not Good' 'Good'
 'Good' 'Not Good']


**Q23.** Count how many 'Good' and 'Not Good' labels there are using `np.unique`.

In [ ]:
labels = np.where(rating > 7, "Good", "Not Good")
names, counts = np.unique(labels, return_counts=True)
print(names)
print(counts)

['Good' 'Not Good']
[481 519]


**Q24.** Sort the `total` array and print its 10 smallest values.

In [ ]:
print(np.sort(total)[:10])

[10.6785 12.6945 13.167  13.419  14.679  16.107  16.2015 16.275  17.094
 18.6375]


**Q25.** Sort the `total` array in descending order and print the top 10 values.

In [ ]:
print(np.sort(total)[::-1][:10])

[1042.65  1039.29  1034.46  1023.75  1022.49  1022.385 1020.705 1003.59
 1002.12   951.825]


`np.sort` only sorts ascending, so add `[::-1]` to flip it.

**Q26.** Use `np.argmax` and `np.argmin` on the `rating` array to print the positions of the highest and lowest rating.

In [ ]:
print("highest rating at position:", np.argmax(rating))
print("lowest rating at position :", np.argmin(rating))

highest rating at position: 60
lowest rating at position : 72


Anything starting with `arg` gives you a position, not a value.

**Q27.** Round the first 10 values of `total` to the nearest whole number.

In [ ]:
print(np.round(total[:10]))

[549.  80. 341. 489. 634. 628. 434. 772.  76. 173.]


**Q28.** Use `np.clip` to cap the `qty` array at a maximum of 8. Print the maximum before and after.

In [ ]:
clipped = np.clip(qty, a_min=None, a_max=8)
print("before:", qty.max())
print("after :", clipped.max())

before: 10
after : 8


**Q29.** Find the cumulative sum of `total` and print its last value.

In [ ]:
cs = np.cumsum(total)
print(cs[-1])

The last value of a cumulative sum is the same as the sum of the whole array.

**Q30.** Find the correlation between `total` and `rating` using `np.corrcoef`.

In [ ]:
print(round(np.corrcoef(total, rating)[0, 1], 4))

The answer is about **-0.04**, which is basically zero. A bigger bill does not mean a happier customer.

---

# Part B — Pandas (30 questions)

Back to the DataFrame `df` from here on.

**Q31.** Print the shape, the column names and the data types of `df`.

In [ ]:
print(df.shape)
print()
print(df.columns)
print()
print(df.dtypes)

**Q32.** Run `df.info()`.

In [ ]:
df.info()

**Q33.** Run `df.describe()`.

In [ ]:
df.describe()

Notice that Total has a mean near 323 but a median near 254. The mean sits higher because a few very large bills pull it up.

**Q34.** Show the first 5 rows and the last 5 rows of the dataset.

In [ ]:
print(df.head())
print()
print(df.tail())

**Q35.** Show 5 random rows of the dataset.

In [ ]:
df.sample(5, random_state=1)

`random_state` keeps the same rows every time you run the cell. Remove it if you want different rows each time.

**Q36.** Print how many unique values each column has.

In [ ]:
print(df.nunique())

**Q37.** Check how many missing values there are in each column.

In [ ]:
print(df.isnull().sum())

Everything is 0 here, so this dataset has no missing values at all.

**Q38.** Make a new DataFrame with only the Branch, City, Product line and Total columns.

In [ ]:
small = df[["Branch", "City", "Product line", "Total"]]
small.head()

Two sets of brackets again. The inner one is the list of column names.

**Q39.** Using `.loc`, show rows 10 to 15 with the columns from City to Quantity.

In [ ]:
df.loc[10:15, "City":"Quantity"]

`.loc` works with labels and it **includes** the last value, so you get 6 rows.

**Q40.** Using `.iloc`, show the last 5 rows and the first 4 columns.

In [ ]:
df.iloc[-5:, :4]

`.iloc` works with positions and it does **not** include the last value.

**Q41.** Show all sales where Total is above 900, and print how many there are.

In [ ]:
big = df[df["Total"] > 900]
print(len(big))
big.head()

**Q42.** Show all sales from the city Yangon.

In [ ]:
yangon = df[df["City"] == "Yangon"]
print(len(yangon))
yangon.head()

**Q43.** Show all sales made by female customers who paid using Ewallet.

In [ ]:
result = df[(df["Gender"] == "Female") & (df["Payment"] == "Ewallet")]
print(len(result))
result.head()

**Q44.** Use `isin` to show all sales from either Yangon or Mandalay.

In [ ]:
result = df[df["City"].isin(["Yangon", "Mandalay"])]
print(len(result))
result.head()

**Q45.** Use `between` to show all sales with a rating between 8 and 10.

In [ ]:
result = df[df["Rating"].between(8, 10)]
print(len(result))
result.head()

**Q46.** Create a new column `Revenue_per_item` equal to Total divided by Quantity.

In [ ]:
df["Revenue_per_item"] = df["Total"] / df["Quantity"]
df[["Total", "Quantity", "Revenue_per_item"]].head()

**Q47.** Create a new column `Rating_Label` that says 'High' when rating is above 7 and 'Low' otherwise.

In [ ]:
df["Rating_Label"] = np.where(df["Rating"] > 7, "High", "Low")
print(df["Rating_Label"].value_counts())

`np.where` works on a pandas column exactly the same way it works on a NumPy array.

**Q48.** Use `.map()` with a dictionary to create a column with the full branch name, for example A becomes 'Branch Alpha'.

In [ ]:
names = {"A": "Branch Alpha", "B": "Branch Beta", "C": "Branch Gamma"}
df["Branch_Name"] = df["Branch"].map(names)
df[["Branch", "Branch_Name"]].head()

**Q49.** Sort the data by Total from highest to lowest and show the top 10 rows.

In [ ]:
df.sort_values("Total", ascending=False).head(10)

**Q50.** Show the 5 sales with the lowest rating.

In [ ]:
df.nsmallest(5, "Rating")

**Q51.** Use `value_counts()` to see how many sales each Product line had. Then show the same thing as a percentage.

In [ ]:
print(df["Product line"].value_counts())
print()
print((df["Product line"].value_counts(normalize=True) * 100).round(2))

**Q52.** Check whether the dataset has any duplicate Invoice ID values.

In [ ]:
print(df["Invoice ID"].duplicated().sum())

The answer is 0, so every bill has its own ID.

**Q53.** Find the total sales for each city and sort from highest to lowest.

In [ ]:
print(df.groupby("City")["Total"].sum().sort_values(ascending=False))

**Q54.** Find the average rating for each Product line.

In [ ]:
print(df.groupby("Product line")["Rating"].mean().round(2))

**Q55.** For each Product line, find the average Rating and the average Total together.

In [ ]:
print(df.groupby("Product line")[["Rating", "Total"]].mean().round(2))

Pass a **list** of columns inside the brackets to get several columns at once.

**Q56.** Group by City and Customer type together, and find the average Total for each combination.

In [ ]:
print(df.groupby(["City", "Customer type"])["Total"].mean().round(2))

**Q57.** Make a pivot table with City as the rows, Product line as the columns and the mean of Total as the values.

In [ ]:
pd.pivot_table(df, index="City", columns="Product line",
               values="Total", aggfunc="mean").round(2)

**Q58.** Make a crosstab of Gender against Payment showing how many sales fall in each combination.

In [ ]:
pd.crosstab(df["Gender"], df["Payment"])

**Q59.** Convert the Date column to datetime, then create a `Month` column and a `DayName` column.

In [ ]:
df["Date"] = pd.to_datetime(df["Date"], format="%m/%d/%Y")
df["Month"] = df["Date"].dt.month
df["DayName"] = df["Date"].dt.day_name()

df[["Date", "Month", "DayName"]].head()

,Date,Month,DayName
0,2019-01-05,1,Saturday
1,2019-03-08,3,Friday
2,2019-03-03,3,Sunday
3,2019-01-27,1,Sunday
4,2019-02-08,2,Friday


Once the column is datetime, `.dt` unlocks month, day, year, day name and a lot more.

**Q60.** Make a table with each city's total sales and average rating, and save it as `sales_summary.csv`.

In [ ]:
summary = df.groupby("City")[["Total", "Rating"]].mean().round(2)
summary.to_csv("sales_summary.csv")

print(summary)
print()
print("Saved as sales_summary.csv")



            Total  Rating
City                     
Mandalay   319.87    6.82
Naypyitaw  337.10    7.07
Yangon     312.35    7.03

Saved as sales_summary.csv


---

# Part C — Seaborn (20 questions)

Give every plot a title using `plt.title()`.

**Q61.** Draw a countplot of Product line.

In [ ]:
plt.figure(figsize=(9, 5))
sns.countplot(data=df, x="Product line")
plt.title("Sales per Product Line")
plt.xticks(rotation=30, ha="right")
plt.tight_layout()
plt.show()

`countplot` counts the rows for you. You do not need to calculate anything before drawing it.

**Q62.** Draw a countplot of Payment.

In [ ]:
plt.figure(figsize=(7, 5))
sns.countplot(data=df, x="Payment")
plt.title("Sales per Payment Method")
plt.tight_layout()
plt.show()

**Q63.** Draw the same countplot of Payment, but with Gender as the hue.

In [ ]:
plt.figure(figsize=(7, 5))
sns.countplot(data=df, x="Payment", hue="Gender")
plt.title("Payment Method by Gender")
plt.tight_layout()
plt.show()

`hue` splits every bar into groups by another column. It works on almost every seaborn plot.

**Q64.** Draw a histogram of Total with 30 bins.

In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df, x="Total", bins=30)
plt.title("Distribution of Bill Amounts")
plt.tight_layout()
plt.show()

**Q65.** Draw a histogram of Rating with a KDE curve on top.

In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df, x="Rating", kde=True)
plt.title("Distribution of Ratings")
plt.tight_layout()
plt.show()

**Q66.** Draw a kdeplot of Total.

In [ ]:
plt.figure(figsize=(8, 5))
sns.kdeplot(data=df, x="Total", fill=True)
plt.title("Bill Amount Curve")
plt.tight_layout()
plt.show()

**Q67.** Draw a boxplot of Total for each Branch.

In [ ]:
plt.figure(figsize=(7, 5))
sns.boxplot(data=df, x="Branch", y="Total")
plt.title("Bill Amounts by Branch")
plt.tight_layout()
plt.show()

**Q68.** Draw a boxplot of Rating for each Product line, and rotate the x labels so they are readable.

In [ ]:
plt.figure(figsize=(10, 5))
sns.boxplot(data=df, x="Product line", y="Rating")
plt.title("Ratings by Product Line")
plt.xticks(rotation=30, ha="right")
plt.tight_layout()
plt.show()

**Q69.** Draw a violinplot of Total for each Customer type.

In [ ]:
plt.figure(figsize=(7, 5))
sns.violinplot(data=df, x="Customer type", y="Total")
plt.title("Bill Amounts: Member vs Normal")
plt.tight_layout()
plt.show()

**Q70.** Draw a stripplot of Rating for each Branch.

In [ ]:
plt.figure(figsize=(8, 5))
sns.stripplot(data=df, x="Branch", y="Rating", alpha=0.5)
plt.title("All Ratings by Branch")
plt.tight_layout()
plt.show()

**Q71.** Draw a scatterplot of Unit price against Total.

In [ ]:
plt.figure(figsize=(8, 6))
sns.scatterplot(data=df, x="Unit price", y="Total")
plt.title("Unit Price vs Total Bill")
plt.tight_layout()
plt.show()

**Q72.** Draw the same scatterplot, but colour the points by Product line.

In [ ]:
plt.figure(figsize=(9, 6))
sns.scatterplot(data=df, x="Unit price", y="Total", hue="Product line")
plt.title("Unit Price vs Total Bill by Product Line")
plt.tight_layout()
plt.show()

**Q73.** Draw a barplot of the average Total for each City.

In [ ]:
plt.figure(figsize=(7, 5))
sns.barplot(data=df, x="City", y="Total")
plt.title("Average Bill by City")
plt.tight_layout()
plt.show()

`barplot` shows the average by default. The thin black line on each bar is the confidence interval.

**Q74.** Draw a barplot of the average Rating for each Product line, with Gender as the hue.

In [ ]:
plt.figure(figsize=(11, 5))
sns.barplot(data=df, x="Product line", y="Rating", hue="Gender")
plt.title("Average Rating by Product Line and Gender")
plt.xticks(rotation=30, ha="right")
plt.tight_layout()
plt.show()

**Q75.** Find the total sales for each Product line and draw a horizontal barplot of the result.

In [ ]:
line_sales = df.groupby("Product line")["Total"].sum().sort_values(ascending=False)

plt.figure(figsize=(9, 5))
sns.barplot(x=line_sales.values, y=line_sales.index)
plt.title("Total Sales by Product Line")
plt.xlabel("Total Sales")
plt.tight_layout()
plt.show()

Putting the long names on the y axis means you do not have to rotate anything.

**Q76.** Draw a heatmap of the correlation between Unit price, Quantity, Total, gross income and Rating, with the numbers shown inside the boxes.

In [ ]:
picked = df[["Unit price", "Quantity", "Total", "gross income", "Rating"]]

plt.figure(figsize=(7, 5))
sns.heatmap(picked.corr(), annot=True, fmt=".2f", cmap="coolwarm")
plt.title("Correlation Heatmap")
plt.tight_layout()
plt.show()

`annot=True` prints the numbers inside the boxes and `fmt='.2f'` keeps them to 2 decimal places.

**Q77.** Group the data by Date, find the total sales per day, and draw a lineplot of it.

In [ ]:
daily = df.groupby("Date")["Total"].sum().reset_index()

plt.figure(figsize=(12, 5))
sns.lineplot(data=daily, x="Date", y="Total")
plt.title("Daily Sales")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

`reset_index()` turns the grouped result back into a normal DataFrame with Date as a proper column, which is what seaborn wants.

**Q78.** Draw a pairplot of Unit price, Quantity and Total.

In [ ]:
sns.pairplot(df[["Unit price", "Quantity", "Total"]])
plt.show()

A pairplot draws every column against every other one, so keep the list short or it gets slow.

**Q79.** Draw a jointplot of Unit price against Total.

In [ ]:
sns.jointplot(data=df, x="Unit price", y="Total", height=6)
plt.show()

`jointplot` builds its own figure, so do not call `plt.figure()` before it.

**Q80.** Draw a catplot of Total by Product line, split into a separate panel for each Branch.

In [ ]:
g = sns.catplot(data=df, x="Product line", y="Total",
                col="Branch", kind="box", height=4.5)

g.set_xticklabels(rotation=60, ha="right")
plt.show()

`col` makes one panel per value of that column. Like jointplot, catplot makes its own figure.

---

## Things worth remembering

- **`.loc` includes the last value, `.iloc` does not.** `df.loc[10:15]` gives 6 rows, `df.iloc[10:15]` gives 5.
- **Use `&` and `|`, not `and` and `or`,** and keep each condition in its own brackets.
- **Two sets of brackets** whenever you pass a list: `total[[0, 10, 50]]` for positions, `df[["City", "Total"]]` for columns.
- **`np.sort` only sorts ascending.** Add `[::-1]` to reverse it.
- **`argmax`, `argmin` and `argsort` return positions, not values.** Anything starting with `arg` gives a position.
- **`axis=0` goes down the rows** and gives one value per column. **`axis=1` goes across the columns** and gives one value per row.
- **`groupby` alone shows nothing.** It only produces a result once you add `.mean()`, `.sum()` or similar.
- **`jointplot`, `pairplot` and `catplot` create their own figure,** so do not call `plt.figure()` before them.